subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -338,44 +342,28 @@ private Object doGet(Type serviceType) throws IllegalArgumentException {
             }
             ServiceProvider provider = providerCache.get(serviceType);
             if (provider == null) {
-                provider = getServiceProvider(serviceType);
+                assertValidServiceType(unwrap(serviceType));
+                DefaultLookupContext context = new DefaultLookupContext();
+                provider = context.find(serviceType, allServices);
+                if (provider == null) {
+                    throw new UnknownServiceException(serviceType, String.format("No service of type %s available in %s.", format(serviceType), getDisplayName()));
+                }
                 providerCache.put(serviceType, provider);
             }
             return provider.get();
         }
     }
 
-    private ServiceProvider getServiceProvider(Type serviceType) {
-        Type lookupType = extractServiceType(serviceType);
-        ServiceProvider provider = hasService(unwrap(lookupType)) ? new DefaultLookupContext().find(serviceType, allServices) : null;
-        if (provider == null) {
-            throw new UnknownServiceException(serviceType, String.format("No service of type %s available in %s.", format(serviceType), getDisplayName()));
-        }
-        return provider;
-    }
-
-    private static Type extractServiceType(Type mayBeFactoryType) {
-        Type serviceType = mayBeFactoryType;
-        if (mayBeFactoryType instanceof ParameterizedType) {
-            Type rawType = ((ParameterizedType) mayBeFactoryType).getRawType();
-            if (rawType == List.class) {
-                serviceType = ((ParameterizedType) mayBeFactoryType).getActualTypeArguments()[0];
-            }
-        }
-        return serviceType;
-    }
-
     public <T> Factory<T> getFactory(Class<T> type) {
         synchronized (lock) {
             if (closed) {
                 throw new IllegalStateException(String.format("Cannot locate factory for objects of type %s, as %s has been closed.", format(type), getDisplayName()));
             }
-            if (hasService(Factory.class)) {
-                DefaultLookupContext context = new DefaultLookupContext();
-                ServiceProvider factory = allServices.getFactory(context, type);
-                if (factory != null) {
-                    return (Factory<T>) factory.get();
-                }
+            assertValidServiceType(Factory.class);
```

Comment:
:x: this is either useless or wrong. Did you mean `assertValidServiceType(type)`?

Reply:
I would say useless, it was like this before. What you propose makes more sense.

Reply:
Fixed